plugins/search-backend-node/src/LunrSearchEngine.ts

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  SearchQuery,
+  IndexableDocument,
+  SearchResultSet,
+} from '@backstage/search-common';
+import lunr from 'lunr';
+import { Logger } from 'winston';
+import { SearchEngine, QueryTranslator } from './types';
+
+type ConcreteLunrQuery = {
+  lunrQueryString: string;
+  documentTypes: string[];
+};
+
+export class LunrSearchEngine implements SearchEngine {
+  protected lunrIndices: Record<string, lunr.Index> = {};
+  protected docStore: Record<string, IndexableDocument>;
+  protected logger: Logger;
+
+  constructor({ logger }: { logger: Logger }) {
+    this.logger = logger;
+    this.docStore = {};
+  }
+
+  translator: QueryTranslator = (query: SearchQuery): ConcreteLunrQuery => {
```

Comment:
Another thought I just had: perhaps our definition of `ConcreteLunrQuery` here is too limited?

Suppose we have two indices: `software-catalog` and `techdocs`.

And suppose we receive a query like this:

```typescript
{
  term: "playlist api",
  filters: {
    namespace: "some-namespace",
    someDocsField: "filter value",
  },
}
```

This might get translated into a `lunrQueryString` like `playlist api namespace:"some-namespace" someDocsField:"filter value"`

...But would result in an error because the `query` method would run the same `lunrQueryString` against both the `software-catalog` and `techdocs` indices.

So maybe we have to open up the definition of `ConcreteLunrQuery` to allow for one `lunrQueryString` per `documentType`?

Reply:
I guess this is tricky for Lunr because the way it handles queries and indices is quite limited...  This will be much more straightforward for more robust search engines, because there is likely already a nicely defined concrete query interface that we can expect.

Reply:
Hmm interesting, so 

> But would result in an error because the query method would run the same lunrQueryString against both the software-catalog and techdocs indices.

do you mean it will fail because "someDocsField" doesn't exist as a field in the document on the software-catalog index? And will it throw and error or just not return any results? because filtering on the "someDocsField" is a way to say lets just search across documents that have that field, so that would be fine I think

Reply:
@iamEAP ^^

Reply:
Yeah, if I'm reading the lunr docs correctly, this section on the [QueryString](https://lunrjs.com/docs/lunr.Index.html#~QueryString) mentions...

> _Using a field not present in the index will lead to an error being thrown._

So as long as we handle that, no worries.  I think your proposal makes sense to (e.g. let the query execute, catch the error, and treat it as no results, since the filter provided would conceptually mean nothing matched anyway)

Reply:
I see, I've handled that + a test for it now!